high-level-caiman/src/normalize/sched_rename.rs

```diff
@@ -0,0 +1,159 @@
+use std::collections::HashMap;
+
+use regex::Regex;
+
+use crate::parse::ast::{
+    ArgsOrEnc, FullType, SchedExpr, SchedFuncCall, SchedLiteral, SchedStmt, SchedTerm,
+};
+
+/// Returns the internal name of a variable, given its original name and id.
+fn internal_name(name: &str, id: u64) -> String {
+    format!("{name}_{id}")
+}
+
+/// Returns the original name of a variable, given its internal name.
+/// For example, `original_name("x_0")` returns `"x"`. If the
+/// name is not an internal name, it is returned unchanged.
+/// # Panics
+/// Regex compilation errors
+#[must_use]
+pub fn original_name(name: &str) -> String {
+    if name.contains(Regex::new(r"_[0-9]+$").unwrap().as_str()) {
```

Comment:
So, I do a simple source-level SSA conversion for internal reasons. Namely to allow shadowing of variables that respects source-level scoping. This conversion essentially appends an underscore followed by a number to *every* variable. And allows something like this:

```
let a = 5;
let b = a + 10;
let a = a + b;
```

Right now (which I just realized is untested) the parser accepts blocks like this:

```
let a = 5;
{
    let b = a;
    let a = b + a;
    let c = a;
}
let d = a + 10;
```

So this whole renaming thing was also to support this type of thing (which doesn't create a new funclet). 

Now in a later PR, I actually ended up applying a more traditional CFG-level SSA transformation after this, which I think makes this earlier transformation unnecessary. But I'd have to think about this more because of some assumptions I might have made in  AST transformations before it's converted to a CFG and the HIR (what I'm calling normalization passes).

Now, this function in particular is solely for error messages so a user sees a variable name they actually used and not some weird mangled version of their variables. Another approach I thought of was to keep a map of some sort mapping new names to old names but then I would need to pass this map around everywhere error messages might start from. So I thought, well, why not just undo the simple transformation I'm applying instead? 

One thing I'm realizing is it doesn't account for references, which are given a special prefix during IR generation, but I wasn't too concerned about handling every case since it's only necessary for error messages and debugging.

Let me know if you have other ideas.

Reply:
Ok, so I did a little more investigation and the main reason I do this renaming is so that variables with the same name in different scopes can have different types.

So for example:
```
if x {
    let b = 10;
} elee {
    let b = true;
}
```

The typing context is global since the unification is done globally, so we don't want these two b's to be thought of as the same metavariable. Hence the renaming so each instance has a different name. 

Now, like I said I do end up applying another SSA renaming later in the next PR to do type deduction of quotients. I was thinking I could remove this first SSA renaming and just use the later version but the problem is that I do base-type deduction at the AST level and quotient deduction at the HIR level. I was debating whether to just do both at the HIR level but the thought I had is that if/when I start introducing more complicated types like tuples, slices, etc. that only exist at high-level-caiman, I'd probably want to do type checking/deduction at this higher level and then lower them to something simpler for the HIR.

All of this is to say, that I think this renaming strategy is required for how things currently work, and seems to be a somewhat reasonable solution (considering the ease of implementing it) to the problem. I will comment this in the code (probably the next PR so I can talk about the interplay between the two renaming passes) since I had to poke around a bit to remember why this.

Then, as for the `original_name` function, I do agree it's a little, ehhh. Here's another idea, instead of names as strings, I could do something like:

```Rust
enum Name {
   Base(String),
   SSA(String, u64)
}
```
Or considering that there's a few renaming passes going on, maybe something like this:

```Rust
enum Adornment {
   SrcSSA(u64),
   Ref,
   HirSSA(u64),
}

struct Name {
   base: String,
   adornments: Vec<Adornment>,
}

// convert Name to string as concatenating base name followed by adornments

// maybe adornments could be sorted by the order in which they are applied
```

I kind of like this last idea and it might result in a simpler way to get the "original name" of a variable from anywhere in the code